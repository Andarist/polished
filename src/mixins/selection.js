// @flow
import messageHandlers from '../internalHelpers/_messageHandlers'

/**
 * CSS to style the selection psuedo-element.
 *
 * ***Deprecation Warning:*** *This mixin has been deprecated and will be removed in version 3.0.*
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...selection({
 *     'backgroundColor': 'blue'
 *   }, 'section')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${selection({'backgroundColor': 'blue'}, 'section')}
 * `
 *
 * // CSS as JS Output
 *
 * 'div': {
 *   'section::-moz-selection': {
 *     'backgroundColor':'blue',
 *   },
 *   'section::selection': {
 *     'backgroundColor': 'blue',
 *   }
 * }
 */

function selection(styles: Object, parent: string = '') {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/selection.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, min: 1, max: 2 },
        typeChecks: [
          { param: styles, type: 'object' },
          { param: parent, type: 'string' },
        ],
      })
    ) {
      return {}
    }
  }

  return {
    [`${parent}::-moz-selection`]: {
      ...styles,
    },
    [`${parent}::selection`]: {
      ...styles,
    },
  }
}

export default selection
