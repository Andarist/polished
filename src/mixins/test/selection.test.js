// @flow
import selection from '../selection'

describe('selection', () => {
  const styles = {
    'background-color': 'blue',
  }

  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should properly pass styles object and parent', () => {
    expect({
      div: {
        ...selection(styles, 'section'),
      },
    }).toMatchSnapshot()
  })

  it('should properly default to no parent when not passed one', () => {
    expect({
      div: {
        ...selection(styles),
      },
    }).toMatchSnapshot()
  })

  it('should throw an error when not passed any parameters', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    selection()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-object as its first parameter', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    selection('styles')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string as its second parameter', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    selection(styles, 1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warning when passed more than 2 parameters', () => {
    selection(styles, 'input', true)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
