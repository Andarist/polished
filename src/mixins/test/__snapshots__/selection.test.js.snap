exports[`selection should properly default to no parent when not passed one 1`] = `
Object {
  "div": Object {
    "::-moz-selection": Object {
      "background-color": "blue",
    },
    "::selection": Object {
      "background-color": "blue",
    },
  },
}
`;

exports[`selection should properly pass styles object and parent 1`] = `
Object {
  "div": Object {
    "section::-moz-selection": Object {
      "background-color": "blue",
    },
    "section::selection": Object {
      "background-color": "blue",
    },
  },
}
`;

exports[`selection should throw a warning when passed more than 2 parameters 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/selection.js --%c

selection %cwill be deprecated as of version 3.0 of ✨ polished. You should use the ::selection pseudo-element instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#selection%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ WARNING ✨ ---- mixins/selection.js --%c

selection %cexpects a maximum of 2 parameters. However, you passed 3 parameters. These additional parameters were ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#selection%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`selection should throw an error when not passed any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/selection.js --%c

selection %cexpects a minimum of 1 parameter. However, you only passed 0 parameters.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#selection%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`selection should throw an error when passed a non-object as its first parameter 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/selection.js --%c

selection %cexpects a 1st parameter of type object. However, you passed styles(string) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#selection%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`selection should throw an error when passed a non-string as its second parameter 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/selection.js --%c

selection %cexpects a 2nd parameter of type string. However, you passed 1(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#selection%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
